cmd/snap/inhibit.go

```diff
@@ -101,76 +135,81 @@ var finishRefreshNotification = func(refreshInfo *client.FinishedSnapRefreshInfo
 	return nil
 }
 
-var tryNotifyRefreshViaSnapDesktopIntegrationFlow = func(snapName string) (bool, error) {
-	// Check if Snapd-Desktop-Integration is available
-	conn, err := dbusutil.SessionBus()
-	if err != nil {
-		logger.Noticef("unable to connect dbus session: %v", err)
-		return false, nil
-	}
-	obj := conn.Object("io.snapcraft.SnapDesktopIntegration", "/io/snapcraft/SnapDesktopIntegration")
-	extraParams := make(map[string]dbus.Variant)
-	err = obj.Call("io.snapcraft.SnapDesktopIntegration.ApplicationIsBeingRefreshed", 0, snapName, runinhibit.HintFile(snapName), extraParams).Store()
-	if err != nil {
-		logger.Noticef("unable to successfully call io.snapcraft.SnapDesktopIntegration.ApplicationIsBeingRefreshed: %v", err)
-		return false, nil
-	}
-	if _, err := waitInhibitUnlock(snapName, runinhibit.HintNotInhibited); err != nil {
-		return false, err
-	}
-	return true, nil
-}
-
-func graphicalSessionFlow(snapName string, hint runinhibit.Hint) error {
-	refreshInfo := client.PendingSnapRefreshInfo{
-		InstanceName: snapName,
-		// Remaining time = 0 results in "Snap .. is refreshing now" message from
-		// usersession agent.
-		TimeRemaining: 0,
-	}
-
-	if err := pendingRefreshNotification(&refreshInfo); err != nil {
-		return err
-	}
-	if _, err := waitInhibitUnlock(snapName, runinhibit.HintNotInhibited); err != nil {
-		return err
+func newInhibitionFlow(snapName string) inhibitionFlow {
+	if isGraphicalSession() {
+		return &graphicalFlow{snapName: snapName}
 	}
-
-	finishRefreshInfo := client.FinishedSnapRefreshInfo{InstanceName: snapName}
-	return finishRefreshNotification(&finishRefreshInfo)
-}
-
-func textFlow(snapName string, hint runinhibit.Hint) error {
-	fmt.Fprintf(Stdout, "%s\n", inhibitMessage(snapName, hint))
-	pb := progress.MakeProgressBar(Stdout)
-	pb.Spin(i18n.G("please wait..."))
-	_, err := waitInhibitUnlock(snapName, runinhibit.HintNotInhibited)
-	pb.Finished()
-	return err
+	return &textFlow{snapName: snapName}
 }
 
-var isLocked = runinhibit.IsLocked
+// waitWhileInhibited blocks until the snap is not inhibited anymore.
+//
+// Calling clean() releases the lock and finishes notification flow.
+//
+// IMPORTANT: It is the callers responsibility to call clean().
+func waitWhileInhibited(snapName string) (clean func() error, retErr error) {
```

Comment:
this is dealing too directly with the lock, so probably it should live as a helper in snaplock/runihnibit taking callbacks to deal with the notification flow etc